service-worker.js

```diff
@@ -23,117 +23,240 @@
+                var headers = new Headers();
+                if (contentLength) headers.set('Content-Length', contentLength);
+                if (contentType) headers.set('Content-Type', contentType);
+                // Test if the content is a video or audio file
+                // See kiwix-js #519 and openzim/zimwriterfs #113 for why we test for invalid types like "mp4" or "webm" (without "video/")
+                // The full list of types produced by zimwriterfs is in https://github.com/openzim/zimwriterfs/blob/master/src/tools.cpp
+                if (contentLength >= 1 && /^(video|audio)|(^|\/)(mp4|webm|og[gmv]|mpeg)$/i.test(contentType)) {
+                    // In case of a video (at least), Chrome and Edge need these HTTP headers or else seeking doesn't work
+                    // (even if we always send all the video content, not the requested range, until the backend supports it)
+                    headers.set('Accept-Ranges', 'bytes');
+                    headers.set('Content-Range', 'bytes 0-' + (contentLength - 1) + '/' + contentLength);
+                }
+                var responseInit = {
+                    status: 200,
+                    statusText: 'OK',
+                    headers: headers
                 };
-                outgoingMessagePort.postMessage({'action': 'askForContent', 'title': titleWithNameSpace}, [messageChannel.port2]);
-            }));
-        }
-        // If event.respondWith() isn't called because this wasn't a request that we want to handle,
-        // then the default request/response behavior will automatically be used.
-    }
+
+                var httpResponse = new Response(msgPortEvent.data.content, responseInit);
+
+                // Let's send the content back from the ServiceWorker
+                resolve(httpResponse);
+            } else if (msgPortEvent.data.action === 'sendRedirect') {
+                resolve(Response.redirect(prefix + msgPortEvent.data.redirectUrl));
+            } else {
+                reject(msgPortEvent.data, titleWithNameSpace);
+            }
+        };
+        outgoingMessagePort.postMessage({
+            'action': 'askForContent',
+            'title': titleWithNameSpace
+        }, [messageChannel.port2]);
+    });
+}
+
+/**
+ * Removes parameters and anchors from a URL
+ * @param {type} url The URL to be processed
+ * @returns {String} The same URL without its parameters and anchors
+ */
+function removeUrlParameters(url) {
+    return url.replace(/([^?#]+)[?#].*$/, '$1');
+}
+
+/**
+ * Looks up a Request in CACHE and returns a Promise for the matched Response
+ * @param {Request} request The Request to fulfill from CACHE
+ * @returns {Response} The cached Response (as a Promise) 
+ */
+function fromCache(request) {
+    // Prevents use of Cache API if user has disabled it
+    if (!useCache) return Promise.reject('disabled');
+    return caches.open(CACHE).then(function (cache) {
```

Comment:
You should test that the Cache API is available before using it, so that it does not break if it's not.
It's probably a rare case, anyway

Reply:
OK, will do.

Reply:
Actually, thinking about this, a test might be redundant as it seems Cache API is a *required part* of the Service Worker spec. According to the [CacheStorage part of the Service Worker spec](https://w3c.github.io/ServiceWorker/#cachestorage-interface), "The user agent _must_ create a CacheStorage object when a Window object or a WorkerGlobalScope object is created and associate it with that global object."  So if we are in Service Worker the Cache API is available, even if we can't actually save a request in it due to having the wrong URL schema. Do you agree @mossroy?

Reply:
You seem to be right. In this case, it's only necessary to test when in app.js